crates/common/trie/state.rs

```diff
@@ -100,4 +100,21 @@ impl TrieState {
         }
         Ok(())
     }
+
+    /// Writes a node batch directly to the DB bypassing the cache
+    pub fn write_node_batch(
+        &mut self,
+        nodes: Vec<Node>,
+        node_hashes: Vec<NodeHash>,
+    ) -> Result<(), TrieError> {
+        // Don't insert the node if it is already inlined on the parent
+        let key_values = node_hashes
+            .into_iter()
+            .zip(nodes)
+            .filter(|(hash, _)| matches!(hash, NodeHash::Hashed(_)))
+            .map(|(hash, node)| (hash.into(), node.encode_to_vec()))
```

Comment:
nit: maybe go to a filter_map instead?

```suggestion
            .filter_map(|(hash, node)| {
                matches!(hash, NodeHash::Hashed(_))
                    .then(|| (hash.into(), node.encode_to_vec()))
            })
```